sonarcloud/resource_project_main_branch.go

```diff
@@ -0,0 +1,189 @@
+				Description: "The key of the project.",
+				Validators: []tfsdk.AttributeValidator{
+					stringLengthBetween(1, 400),
+				},
+			},
+		},
+	}, nil
+}
+
+func (r resourceProjectMainBranchType) NewResource(_ context.Context, p tfsdk.Provider) (tfsdk.Resource, diag.Diagnostics) {
+	return resourceProject{
+		p: *(p.(*provider)),
+	}, nil
+}
+
+type resourceProjectMainBranch struct {
+	p provider
+}
+
+func (r resourceProjectMainBranch) Create(ctx context.Context, req tfsdk.CreateResourceRequest, resp *tfsdk.CreateResourceResponse) {
+	if !r.p.configured {
+		resp.Diagnostics.AddError(
+			"Provider not configured",
+			"The provider hasn't been configured before apply, likely because it depends on an unknown value from another resource. "+
+				"This leads to weird stuff happening, so we'd prefer if you didn't do that. Thanks!",
+		)
+		return
+	}
+
+	// Retrieve values from plan
+	var plan ProjectMainBranch
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	request := project_branches.RenameRequest{
+		Project: plan.ProjectKey.Value,
+		Name:    plan.Name.Value,
+	}
+
+	err := r.p.client.ProjectBranches.Rename(request)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Could not create the main project branch",
+			fmt.Sprintf("The Rename request returned an error: %+v", err),
+		)
+		return
+	}
+}
+
+func (r resourceProjectMainBranch) Read(ctx context.Context, req tfsdk.ReadResourceRequest, resp *tfsdk.ReadResourceResponse) {
+	// Retrieve values from state
+	var state ProjectMainBranch
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Fill in api action struct
+	request := project_branches.ListRequest{
+		Project: state.ProjectKey.Value,
+	}
+
+	response, err := r.p.client.ProjectBranches.List(request)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Could not read the project branches",
+			fmt.Sprintf("The List request returned an error: %+v", err),
+		)
+		return
+	}
+
+	// Check if the main branch matches the declared main branch
+	if result, ok := findProjectMainBranch(response, state.Name.Value, state.ProjectKey.Value); ok {
+		diags = resp.State.Set(ctx, result)
+		resp.Diagnostics.Append(diags...)
+	} else {
+		resp.State.RemoveResource(ctx)
+	}
+}
+
+func (r resourceProjectMainBranch) Update(ctx context.Context, req tfsdk.UpdateResourceRequest, resp *tfsdk.UpdateResourceResponse) {
+	// Retrieve values from state
+	var state ProjectMainBranch
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Retrieve values from plan
+	var plan ProjectMainBranch
+	diags = req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	changed := changedAttrs(req, diags)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	projectKey := state.ProjectKey.Value
```

Comment:
Nitpick: if changing the `project_key` causes a destroy (see my other comment regarding which attribute should semantically trigger a destroy), then it should never show up as changed in an update. It should be safe to remove this.